chalk-recursive/src/recursive.rs

```diff
@@ -223,24 +211,15 @@ impl<'me, I: Interner> SolveDatabase<I> for Solver<'me, I> {
         if let Some(dfn) = self.context.search_graph.lookup(&goal) {
             // Check if this table is still on the stack.
             if let Some(depth) = self.context.search_graph[dfn].stack_depth {
-                // Is this a coinductive goal? If so, that is success,
-                // so we can return normally. Note that this return is
-                // not tabled.
-                //
-                // XXX how does caching with coinduction work?
-                if self.context.stack.coinductive_cycle_from(depth) {
-                    let value = ConstrainedSubst {
-                        subst: goal.trivial_substitution(self.program.interner()),
-                        constraints: Constraints::empty(self.program.interner()),
-                    };
-                    debug!("applying coinductive semantics");
-                    return Ok(Solution::Unique(Canonical {
-                        value,
-                        binders: goal.canonical.binders,
-                    }));
-                }
-
                 self.context.stack[depth].flag_cycle();
+                // Mixed cycles are not allowed.
```

Comment:
I'm currently skimming through [this paper](https://link.springer.com/chapter/10.1007%2F978-3-540-73420-8_42) which is based on the one you mentioned but concentrates on combining inductive and coinductive semantics. Combined with your explanation it should help me formulating the section for the Chalk book.

Reply:
I did a bit of research yesterday and it seems that there exist ways to handle mixed cycles by extending the logic (cf. [this paper](https://arxiv.org/abs/2008.02140)). Though, I'm not sure yet, whether this would be possible for Chalk.